### PR TITLE
Update Android in CI

### PR DESCRIPTION
**Pull request by maurer**
We are currently using a 10+ year old Android image, and it has caused trouble when working on #120326.

Our current NDK (25) only supports API 19+, so we were already out of spec. This PR:

1. Bumps the API used by the emulator in CI to 21, as per [NDK-26's release notes](https://github.com/android/ndk/wiki/Changelog-r26) deprecating 19 and 20 as targets.
2. Bumps the NDK to 26b

try-job: arm-android

### CODE REVIEW

**Comment by maurer** [comment]
r? @pietroalbini 

Pietro, I've added you explicitly since you're the only one to ever touch this lockfile, which means you're likely to be the one who can run `update-mirror`, which I lack the credentials for.

**Comment by maurer** [comment]
cc @chriswailes

**Comment by Mark-Simulacrum** [comment]
The new files should be uploaded to the mirror. However, I'd like to get a +1 from another android target maintainer on this PR before moving ahead.

cc @chriswailes @mgeisler (per [platform support docs](https://doc.rust-lang.org/nightly/rustc/platform-support/android.html))

**Review by chriswailes — APPROVED** [review]
Lgtm!

**Comment by workingjubilee** [comment]
Happy to see this land, as it looks like it will also allow simplifying away https://github.com/rust-lang/backtrace-rs/pull/570 from backtrace-rs, too.

**Comment by workingjubilee** [comment]
Failed about here, it seems:

```
 Executing python3 ../x.py dist --host= --target arm-linux-androideabi,armv7-linux-androideabi,thumbv7neon-linux-androideabi,i686-linux-android,aarch64-linux-android,x86_64-linux-android
+ python3 ../x.py dist --host= --target arm-linux-androideabi,armv7-linux-androideabi,thumbv7neon-linux-androideabi,i686-linux-android,aarch64-linux-android,x86_64-linux-android
Building bootstrap
      Finished `dev` profile [unoptimized] target(s) in 0.04s
thread 'main' panicked at src/core/sanity.rs:59:13:


couldn't find required command: "/android/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi19-clang"


note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by maurer** [comment]
I missed the [`cc_detect`](https://github.com/rust-lang/rust/pull/120593/commits/93acd1bee3abe8f5eef6cf169c255cf9ac576a22#diff-cfa9af18d2fd250261e30a58dddc62b08a95dc830da96be17a1849730d9b44e1) logic the first time around - it was still attempting to pick an API 19 linker for 32-bit.

@rustbot ready

**Comment by workingjubilee** [comment]
The change in `cc_detect` caused me to inspect and notice that config.toml.example had a ref to n25b so I asked maurer to update that, which led to the etc. etc. There are some other necessary followups too, but this CI change should land independently.

**Comment by workingjubilee** [comment]
The fact this timeout was on, specifically, the aarch64-android job, makes me feel like this was not a spurious error, but I have no proof for that.

**Comment by maurer** [comment]
Seems possible that it's a real failure. Are you aware of any workflows for trying to reproduce locally? I would have expected that even a timeout would have logs leading up to the failing step...

**Comment by Mark-Simulacrum** [comment]
https://rustc-dev-guide.rust-lang.org/tests/docker.html

https://github.com/rust-lang-ci/rust/actions/runs/8411885535/job/23032016177 for the logs.

**Comment by workingjubilee** [comment]
The logs just terminate somewhat cryptically here:

```
Building stage2 library artifacts (x86_64-unknown-linux-gnu -> arm-linux-androideabi)
[TIMING] core::build_steps::compile::Std { target: arm-linux-androideabi, compiler: Compiler { stage: 2, host: x86_64-unknown-linux-gnu }, crates: [], force_recompile: false, extra_rust_args: [], is_for_mir_opt_tests: false } -- 46.623
REMOTE copy libs to emulator (arm-linux-androideabi)
Building stage2 tool remote-test-server (x86_64-unknown-linux-gnu -> arm-linux-androideabi)
     Compiling remote-test-server v0.1.0 (/checkout/src/tools/remote-test-server)
  [RUSTC-TIMING] remote_test_server test:false 0.682
      Finished `release` profile [optimized] target(s) in 0.85s
[TIMING] core::build_steps::tool::ToolBuild { compiler: Compiler { stage: 2, host: x86_64-unknown-linux-gnu }, target: arm-linux-androideabi, tool: "remote-test-server", path: "src/tools/remote-test-server", mode: ToolStd, source_type: InTree, extra_features: [], allow_features: "" } -- 0.864
[TIMING] core::build_steps::tool::RemoteTestServer { compiler: Compiler { stage: 2, host: x86_64-unknown-linux-gnu }, target: arm-linux-androideabi } -- 0.001
Building stage0 tool remote-test-client (x86_64-unknown-linux-gnu)
     Compiling remote-test-client v0.1.0 (/checkout/src/tools/remote-test-client)
  [RUSTC-TIMING] remote_test_client test:false 0.579
      Finished `release` profile [optimized] target(s) in 0.73s
[TIMING] core::build_steps::tool::ToolBuild { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu }, target: x86_64-unknown-linux-gnu, tool: "remote-test-client", path: "src/tools/remote-test-client", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "" } -- 0.746
[TIMING] core::build_steps::tool::RemoteTestClient { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu }, target: x86_64-unknown-linux-gnu } -- 0.001
waiting for device to come online
* daemon not running; starting now at tcp:5037
* daemon started successfully
Error: The operation was canceled.
```

The log analyzer decided not to report it because its heuristics meant this was decided to not be very informative, which... is correct, honestly. With a "wonderful" log report like that, it's hard to actually rule out anything. Though, it probably should have decided to just fork over those last few lines anyways.

**Comment by Dylan-DPC** [comment]
@maurer any updates on this? thanks

**Comment by maurer** [comment]
Sorry, I've been on vacation for the last two weeks. I should have a chance to investigate this further some time this week.

**Comment by maurer** [comment]
Brief update:
1. We're going to need to upload `emulator emulator-linux_x64-11772612.zip 974ad0591834a9fa3db307e49cdba352462e5f1e` - the emulator that was in the previous change has a bug when running in headless mode that leads to segfaults/crashes
2. The `aarch64` development image does not support 32-bit execution. We'll need a separate `system-images;android-21;default;armeabi-v7a`.

Would folks prefer:
a. Updating the android builder to launch two VMs, and do both arm32 and aarch64 testing on them
b. Split into two builders, and keep to the 'one external runner per builder' paradigm that's been in use.

As far as I can tell, b is closer to what we've done elsewhere, so it's what I'll configure if I don't hear otherwise. I'll update the change once I have something passing (with the caveat of a hack I'll need until the mirror is updated).

**Comment by maurer** [comment]
@Mark-Simulacrum , since you uploaded the files last time (so I know you have the power), can you update with *both* `.lock` files (now that I created two of them, with different contents)? Once that's done, I'll pull the two hack commits I've got on for testing off and we should finally be good.

**Comment by Mark-Simulacrum** [comment]
Happy to upload some files to our mirror, but I'm not sure where I should mirror them from or exactly which files need to be uploaded. Can you clarify?

**Comment by maurer** [comment]
Specifically, I would like you to run:
```
src/ci/docker/scripts/android-sdk-manager.py update-mirror src/ci/docker/host-x86_64/aarch64-android/android-sdk.lock
src/ci/docker/scripts/android-sdk-manager.py update-mirror src/ci/docker/host-x86_64/arm-android/android-sdk.lock
```
with this CL pulled down.

This will fetch the Android images + emulator from google's repositories and upload them to your buckets. I'll then drop [DO NOT SUBMIT: Hack SDK manager to not use buckets](https://github.com/rust-lang/rust/pull/120593/commits/5106c53e11f0bb09f8881bf9fb0d3549bc0d2bc2) from the PR to make sure it worked (let GHA run), and then drop [DO NOT SUBMIT - force run aarch64-android and arm-android](https://github.com/rust-lang/rust/pull/120593/commits/8da3068c13871da5482ca87eeb8e7e5d740750d0) and it should be good for submission.

**Comment by Mark-Simulacrum** [comment]
Alright, that should be done.

**Comment by maurer** [comment]
OK, ready for review.
[aarch64 run](https://github.com/rust-lang/rust/actions/runs/9044831438/job/24853916111?pr=120593)
[arm run](https://github.com/rust-lang/rust/actions/runs/9044831438/job/24853916027?pr=120593)

**Comment by workingjubilee** [comment]
@maurer sorry about that, can you rebase this?

**Comment by Mark-Simulacrum** [comment]
@bors retry 

aarch64 runner reboot

**Comment by maurer** [comment]
That looks like a pointer getting used as an index - that number is `0xffffff8aee5d1001`. I don't think it's related to anything in this CL, but I can try re-running things manually tomorrow...

**Comment by workingjubilee** [comment]
hmm... this is a really normal-looking test. it's not obvious as to what would be causing this. https://github.com/rust-lang/rust/blob/606afbb617a2949a4e35c4b0258ff94c980b9451/library/alloc/src/collections/vec_deque/tests.rs#L656-L680

**Comment by maurer** [comment]
Can I get a CI run, and if the CI run succeeds, get a re-upload of the lockfile contents? It looks like we did in fact just have a busted emulator.

**Comment by Kobzol** [comment]
I added try-job annotations to the PR description, so that we can test the Android CI jobs in try builds without modifying the CI config.

**Comment by maurer** [comment]
Moving it into the try-spec failed because my earlier hack did it by moving them *out* of `auto`. I've put them back.

Note that I still require one hack commit for tests to avoid using the S3 cache bucket until someone from the team has a chance to update it.

**Comment by workingjubilee** [comment]
@maurer 

Uhh, hasn't that already been uploaded since https://github.com/rust-lang/rust/pull/120593#issuecomment-2105932019

**Comment by maurer** [comment]
> @maurer
> 
> Uhh, hasn't that already been uploaded since [#120593 (comment)](https://github.com/rust-lang/rust/pull/120593#issuecomment-2105932019)

That uploaded the broken emulator, unfortunately. The change between then and now is literally to grab a different emulator. I can take off the hack commit if you want to see what happens.

**Comment by workingjubilee** [comment]
oh okay.

**Comment by maurer** [comment]
New error, but equally odd to the old one.

My build was a bit out of date, so I'm rebasing locally and rerunning to see if I can replicate...

**Comment by maurer** [comment]
Locally, I'm getting
```
thread 'io::tests::test_write_all_vectored' panicked at library/core/src/panicking.rs:221:5:
unsafe precondition(s) violated: slice::from_raw_parts requires the pointer to be aligned and non-null, and the total size of the slice not to exceed `isize::MAX`
thread caused non-unwinding panic. aborting.
```
so it looks like something is still wrong. Maybe I had a lucky pass before? I'll try to investigate.

**Comment by maurer** [comment]
So, this is just a rebase relative to the previous commit, but `./src/ci/docker/run.sh aarch64-android` has passed twice locally.

Can I get a `@bors try` to see if this has disappeared on CI, since I can't seem to reproduce anymore?

**Comment by tgross35** [comment]
Seems like this might have lost its workflows in the outage

**Comment by maurer** [comment]
Dummy commit stripped off since the tryjob passed.

Before merging/approving, you'll need to re-upload the aarch64-android binaries. The specific command is:
```
src/ci/docker/scripts/android-sdk-manager.py update-mirror src/ci/docker/host-x86_64/aarch64-android/android-sdk.lock
```

@rustbot ready

**Comment by Mark-Simulacrum** [comment]
Ran the script.

@bors try

**Comment by workingjubilee** [comment]
Let's try this bad boy against our full CI.

@bors r=MarkSimulacrum,workingjubilee rollup=NEVER

**Comment by maurer** [comment]
... it passed via try twice, and locally via `src/ci/docker/run.sh aarch64-android` twice. Guess back to the drawing board with trying to replicate this :(

**Comment by tgross35** [comment]
Appears CI got stuck shortly after `core::build_steps::tool::RemoteTestClient`. Seems unlikely to be caused by the changes since we have seen this job pass.

@bors try-

**Comment by tgross35** [comment]
Hmm doesn't seem to have picked up the cancel, but maybe the new job will override it?

@bors try

**Comment by tgross35** [comment]
Seems like it is stuck at the same point, which appears to be just before "pushing server" based on logs from https://github.com/rust-lang-ci/rust/actions/runs/10101022677/job/27933675737. @maurer any ideas here?

**Comment by maurer** [comment]
That would normally mean that the virtual device wasn't coming up, so it was hanging on connecting... but wasn't this working before?

**Comment by tgross35** [comment]
It did pass at https://github.com/rust-lang-ci/rust/actions/runs/10049405628/job/27775374879. The dropped patch 005ab822d903f76342fe55de2a867d8b1486574e ("Add aarch64-android testing") included a change to the arm dockerfile, as well as android-start-emulator.sh - any chance that could be it?

**Comment by maurer** [comment]
Ah, I dropped that because I didn't need to specify the arch anymore, but specifying the `classic` engine might have been load-bearing.

**Comment by tgross35** [comment]
Great, time for the real thing. Only change from the previous approval was to drop the patches that add aarch64.

@bors r=Mark-Simulacrum,workingjubilee

**Comment by tgross35** [comment]
> The job **`arm-android`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10102411404/job/27937948493) [(plain)](https://github.com/rust-lang-ci/rust/commit/47775ff3f9460ec4b3835646dc0570286ad3fbc5/checks/27937948493/logs)
> Click to see the possible cause of the failure (guessed by this bot)

(this is just the job that started 4 hours ago timing out)

**Comment by Kobzol** [comment]
> Hmm doesn't seem to have picked up the cancel, but maybe the new job will override it?

Note that try builds cannot be cancelled using the current bors bot.

**Comment by danakj** [comment]
Congratulations on getting this merged, it seems to have been incredibly difficult. Thanks for the persistence.

**Comment by briansmith** [comment]
> Bumps the API used by the emulator in CI to 21, as per [NDK-26's release notes](https://github.com/android/ndk/wiki/Changelog-r26) deprecating 19 and 20 as targets.


This implies that API level 21 is the new minimum supported (at least minimum *tested*) API level, which should be documented...somewhere, especially somewhere in (or linked from) the platform support documentation.

> Bumps the NDK to 26b

Ditto. This seems to imply the minimum NDK version is 26b now, which should be similarly documented.

**Comment by workingjubilee** [comment]
I think the actually germane element here is the API level. If you compile Rust with the oldest NDK that supports our minimum API level, it *should* work, in theory?

Not that we're testing it, so we should be clear that that's the NDK we *support*, but I draw the distinction because the API level is the **hard** limit.

**Comment by maurer** [comment]
In a [blog post](https://blog.rust-lang.org/2023/01/09/android-ndk-update-r25.html) for 1.68, we noted that we were moving to a policy of targeting the most recent LTS NDK. This is further documented in the [NDK/API Update Policy](https://doc.rust-lang.org/rustc/platform-support/android.html#ndkapi-update-policy) under Platform Support.

I'd agree with you that it would be worth an extra notification if we were bumping the API beyond that, but we're bumping the min API exactly the way we messaged we would - we currently support all API levels supported by the supported NDK, which is the LTS NDK, and results in a minimum of 21.

**Comment by briansmith** [comment]
Do we still intend to support API level 19 even though we're not testing it? Or are we now allowing the use of API level 20/21 features in future commits? This is the kind of clarification that I would like to see.

In my experience with MSRV, soon after I stop testing with the MSRV toolchain, the build with that toolchain breaks. More generally, "supported but not tested" minimum supported toolchain versions tend not to be "supported" very well.

**Comment by workingjubilee** [comment]
@briansmith No? And I'm not sure how you apprehended that conclusion.

I was only commenting about the API/NDK distinction because older NDKs can theoretically compile for that API level. The API level still represents the hard minimum. The API level is what defines the actual implementation of the standard library.

**Pull request merged**
